### ISSUE TITLE
git machete changes `.git/rebase-merge/author-script` to windows-style newlines, breaks rebasing

### ISSUE THREAD

**Issue by cspotcode** [PawelLipski/git-machete#935]
# Overview

On windows, git-machete converts the newlines in `.git/rebase-apply/author-script` into `\r\n`, but git requires them to be `\n`.  Git thus raises an error when rebasing.

https://github.com/VirtusLab/git-machete/blob/846ca1e6b4266b507b01e1b928ef5dadc882ca11/git_machete/git_operations.py#L941

Apparently python normalized newlines to the OS (`\r\n` on windows) unless you take care to disable that behavior.

One way to do that is with `io.open()`

```python
io.open('foo.txt', 'w', newline='').write('foo\nbar\nbiff\n')
```

## Specific scenario when it occurs

When running a git machete command that causes a rebase to happen, and there are conflicts in the rebase.  Then git drops me back to my terminal, I address the conflict, `git add <conflicted file>`.  Then, when I run` git commit`, I get an error that git cannot parse `.git/rebase-apply/author-script`

## Expected behavior/command output

Rebasing works on windows.

## Actual behavior/command output

Aforementioned error.

## `git` and `git machete` versions

```
git --version
git version 2.40.1.windows.1
git machete --version
git machete version 3.17.1
```

## Please `cp -r .git .git.bak` or there's no hope!

Please back-up your `.git` folder at its current state for the issue to be reproducible in the future!
Since `git machete` heavily relies on branch reflogs (`.git/logs/refs/`) that change as state of the repository changes, get expired due to `git gc` etc., not doing this almost guarantees we won't be able to reproduce the issue and determine whether it was fixed or not.

## Additional diagnostics info

Depending on your NDAs/level of confidentiality of the project stored in your repository, you can also include **one of the 3 following variants** of diagnostics information:

1. `--verbose` - executed git commands

    `git machete --verbose <problematic-command>` (stdout+stderr)

    If it comes to repository-specific data, this will only include names of local branches and their remote counterparts. No commit messages, file paths or file contents will be included.

2. `--debug` - above + git command outputs and other detailed logging information

    `git machete --debug <problematic-command>` (stdout+stderr)

    If it comes to repository-specific data, this will include names of local branches and their remote counterparts, their full logs (but only wrt. commit hashes) and their full reflogs (which typically include commit messages). No file paths or file contents will be included.

3. Ultimate knowledge - above + the `.git` folder

    `git machete --debug <problematic-command>` (stdout+stderr)

    AND

    `tar czvf dotgit.tgz .git/`

    Sharing contents of `.git` will provide most detailed diagnostics info, but also obviously involves sharing the entire local history of the repository. Mostly recommended for open source projects.

    In case the tgz-ed repository is too large to share via GitHub web upload, you can try either pushing the tgz-ed repository as a file onto a branch, or actually sharing the entire non-tgz-ed `.git` folder as a repository on GitHub, so basically doing:

    ```shell script
    cd .git/
    git init
    git add -a
    git commit -m 'git-machete v?.?.? issue #...'
    git remote add origin ...URL...
    git push origin master
    ```

Obviously sharing more data will make it easier to investigate the issue, but also involves sharing more project data.

**Comment by PawelLipski** [comment]
To be fixed soon! 🔍

**Comment by PawelLipski** [comment]
Huh there's one catch... git-machete never writes `.git/rebase-apply/author-script`, only `.git/rebase-merge/author-script` ‼️   

https://github.com/VirtusLab/git-machete/blob/846ca1e6b4266b507b01e1b928ef5dadc882ca11/git_machete/git_operations.py#L926-L929

The rebase that fails... is it interactive or non-interactive (I'd assume the latter since `rebase-apply` is involved)?
Does a rebase, executed in that mode **outside** of git-machete, succeed without the `\n` problem?

**Comment by cspotcode** [comment]
I may have gotten the two filenames mixed up since I submitted the bug report some time after I'd worked around the problem.  I remember opening the file and changing it from crlf to lf, which fixed it.  I had been using traverse and slide-out commands.

**Comment by PawelLipski** [comment]
Okay, then I'll provide the fix assuming that `.git/rebase-merge/author-script` was affected

**Comment by PawelLipski** [comment]
> On windows, git-machete converts the newlines in .git/rebase-apply/author-script into \r\n, but git requires them to be \n

I'm just trying to reproduce the error... I've cooked up a [test](https://github.com/VirtusLab/git-machete/pull/942/files#) and [executed it on a Windows machine](https://app.circleci.com/pipelines/github/VirtusLab/git-machete/3619/workflows/33d3f129-bc74-4365-b26a-4f2cd80f2546/jobs/19618). I can indeed see that `\r\n` is written to `author-script`:

```xxd
00000000: 4749 545f 4155 5448 4f52 5f4e 414d 453d  GIT_AUTHOR_NAME=
00000010: 2754 6573 7465 7220 5465 7374 270d 0a47  'Tester Test'..G
00000020: 4954 5f41 5554 484f 525f 454d 4149 4c3d  IT_AUTHOR_EMAIL=
00000030: 2774 6573 7465 7240 7465 7374 2e63 6f6d  'tester@test.com
00000040: 270d 0a47 4954 5f41 5554 484f 525f 4441  '..GIT_AUTHOR_DA
00000050: 5445 3d27 4031 3638 3338 3135 3232 3420  TE='@1683815224 
00000060: 2b30 3030 3027 0d0a                      +0000'..
```

but `git rebase --continue` succeeds regardless afterwards 🤔 So...

> Git thus raises an error when rebasing.

Can you dig out the exact error? from history or just repro the situation 🤔

**Comment by cspotcode** [comment]
I figured out a repro.

My commit graph looks like this:
<img width="345" alt="image" src="https://github.com/VirtusLab/git-machete/assets/376504/1a5c734c-7afd-4253-a1af-223e9e9e94c8">

My git machete config file looks like this:
```
main
    feature-1
        feature-2
```

And then I do `git machete slide-out feature-1`

```
repro> git machete slide-out feature-1
Rebasing feature-2 onto main...
Auto-merging README.md
CONFLICT (content): Merge conflict in README.md
error: could not apply c96113b... change on feature-2
hint: Resolve all conflicts manually, mark them as resolved with
hint: "git add/rm <conflicted_files>", then run "git rebase --continue".
hint: You can instead skip this commit: run "git rebase --skip".
hint: To abort and get back to the state before "git rebase", run "git rebase --abort".
Could not apply c96113b... change on feature-2
git rebase --interactive --empty=drop --onto refs/heads/main c24b503c83ae59bd89a676d3b8b7f5ccc2a3bbc4 feature-2 returned 1

```
It attempts to rebase a single commit from `feature-2` onto `main` but the commit has a confict in `README.md`.  I resolve the conflict and continue the rebase.

```
repro> git add .\README.md
repro> git rebase --continue    
error: unable to dequote value of 'GIT_AUTHOR_NAME'
error: you have staged changes in your working tree
If these changes are meant to be squashed into the previous commit, run:

  git commit --amend 

If they are meant to go into a new commit, run:

  git commit 

In both cases, once you're done, continue with:

  git rebase --continue

error: could not commit staged changes.
```

`.\.git\rebase-merge\author-script` has CRLF line endings.  I change them to LF, safe, re-try `git rebase --continue` and it succeeds.

**Comment by PawelLipski** [comment]
Ha! I've got a [repro](https://app.circleci.com/pipelines/github/VirtusLab/git-machete/3621/workflows/4ba12a3f-e8a6-4779-a5fe-6d20a4ad8f1e/jobs/19620) now, thanks 🎉 the missing piece was that some file needed to be staged when `git rebase --continue` is executed. Fix on the way....

**Comment by PawelLipski** [comment]
Fixed! v3.17.4 will be released soon
